components/data_stores/actions/get-record-keys/get-record-keys.mjs

```diff
@@ -0,0 +1,114 @@
+import app from "../../data_stores.app.mjs";
+
+export default {
+  key: "data_stores-get-record-keys",
+  name: "Get Record Keys",
+  description: "Get all record keys in your [Pipedream Data Store](https://pipedream.com/data-stores/) that mathces with your query. The memory consumption of the workflow can be affected, since this action will be exposing, to the workflow, the entire data from the selected datastore",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    app,
+    dataStore: {
+      propDefinition: [
+        app,
+        "dataStore",
+      ],
+    },
+    query: {
+      type: "string",
+      label: "Query",
+      description: "The query to filter the records.",
+    },
+    contains: {
+      type: "boolean",
+      label: "Contains",
+      description: "If true, the query will be used to filter the records that contains the query. If false, the query will be used to filter the records that are equal to the query.",
+      optional: true,
+    },
+    caseInsensitive: {
+      type: "boolean",
+      label: "Case Insensitive",
+      description: "if true, the query will be used to filter the records based on case-Insensitivity, if false, the query will be used to filter the records that are equal to the query.",
+      optional: true,
+    },
+  },
+  methods: {
+    getAllKeysWithRecord(
+      obj,
+      query,
+      contains = false,
+      caseInsensitive = false,
+      path = "",
+    ) {
+      const filteredKeys = [];
+
+      const helper = (obj, path) => {
+        if (Array.isArray(obj)) {
+          for (let i = 0; i < obj.length; i++) {
+            helper(obj[i], path + i + ".");
+          }
+          return;
+        }
+        else if (typeof obj === "object" && obj !== null) {
+          for (let key in obj) {
+            helper(obj[key], path + key + ".");
+          }
+          return;
+        }
+
+        if (!contains && !caseInsensitive && obj == query) {
+          filteredKeys.push(path.slice(0, -1));
+          return;
+        }
+
+        if (contains && caseInsensitive && String(obj).toLowerCase()
+          .includes(query.toLowerCase())
+        ) {
+          filteredKeys.push(path.slice(0, -1));
+          return;
+        }
+
+        if (contains && !caseInsensitive && String(obj).includes(query)) {
+          filteredKeys.push(path.slice(0, -1));
+          return;
+        }
+
+        if (!contains && caseInsensitive
+          && String(obj).toLocaleLowerCase() == query.toLowerCase()
+        ) {
+          filteredKeys.push(path.slice(0, -1));
+          return;
+        }
```

Comment:
I find this section of the code somewhat challenging to understand.

There are numerous base cases for the recursive portion, which can be a bit overwhelming. Additionally, the second part contains an excessive amount of conditions.

I am also not entirely sure of the logic behind the combination of the path and the index.

Furthermore, it appears that some of the code is repeated, which could potentially be improved.

Lastly, I'm not sure if `query` is being included in the scope of the helper function.

Reply:
There are other ways to make the code more readable. For example, create methods:
```js
containsQuery(obj, query, caseSensitive = false) {
  obj = String(obj);
  return caseSensitive
    ? obj.includes(query)
    : obj.toLowerCase().includes(query.toLowerCase());
}

matchesQuery(obj, query, caseSensitive = false) {
  obj = String(obj);
  return caseSensitive
    ? obj === query
    : obj.toLowerCase() === query.toLowerCase();
}
```

And then use the returned `boolean` value to
```js
if (matched) {
  filteredKeys.push(path.slice(0, -1));
  values.push(obj);
}
```

Reply:
This is just my preference but for me it is clearer to use `caseSensitive` instead of `caseInsensitive`. Especially when having to negate it twice, e.g. `!caseInsensitive`.